File: Data Structures & Algorithms/three-integer-sum/submission-11.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());

        for(int i = 0; i < nums.size(); ++i){
            if(i > 0 && nums[i] == nums[i-1]){
                continue;
            }

            int l = i+1;
            int r = nums.size()-1;

            while(l < r){
                int total = nums[i] + nums[l] + nums[r];
                if(total == 0){
                    res.push_back({nums[i], nums[l], nums[r]});
                    ++l;
                    while(l < r && nums[l] == nums[l-1]){
                        ++l;
                    }
                }else if(total > 0){
                    --r;
                }else{
                    ++l;
                }
            }

        }

        return res;

    }
};

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-11.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l, r = 0, len(nums)-1
        res = float('inf')

        # loop through nums
        while l <= r:
            m = (l+r) // 2

            # Check if sorted
            if nums[l] <= nums[r]:
                res = min(res, nums[l])


            # check if on left or right side
            
            # if on rightside, save then move left one
            if nums[l] > nums[m]:
                res = min(res, nums[m])
                r = m - 1
            else:
                # if on left side, move to right
                l = m + 1

        return res